pynamodb/attributes.py

```diff
@@ -945,18 +943,41 @@ def deserialize(self, values):
         """
         Decode from list of AttributeValue types.
         """
-        deserialized_lst = []
-        for v in values:
-            attr_type, attr_value = next(iter(v.items()))
-            attr_class = self._get_deserialize_class(attr_type)
-            if attr_class.attr_type != attr_type:
-                raise ValueError("Cannot deserialize {} elements from type: {}".format(
-                    attr_class.__class__.__name__, attr_type))
-            deserialized_lst.append(attr_class.deserialize(attr_value))
-        return deserialized_lst
-
-    def __getitem__(self, idx: int) -> Path:
-        # for typing only
+        if self.element_type:
+            element_attr = self.element_type()
+            if isinstance(element_attr, MapAttribute):
+                element_attr._make_attribute()  # ensure attr_name exists
+            deserialized_lst = []
+            for idx, attribute_value in enumerate(values):
+                value = None
+                if NULL not in attribute_value:
+                    # set attr_name in case `get_value` raises an exception
+                    element_attr.attr_name = '{}[{}]'.format(self.attr_name, idx)
+                    value = element_attr.deserialize(element_attr.get_value(attribute_value))
```

Comment:
Additionally this change modifies `deserialize` to call `get_value` so that attributes which special case attribute lookup (e.g. for migrations) are properly supported in lists.